### PR TITLE
Feat: Dependent features

### PR DESCRIPTION
**Pull request by RikudouSage**
# Description

Adds support for dependent features.

Replaces #176 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Unit tests
- [x] Spec Tests
- [x] Integration tests / Manual Tests

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/DefaultUnleashTest.php [hunk 1]
```diff
@@ -813,6 +826,182 @@ public function testAutoRegistrationOnInvalidResponse()
         );
     }
 
+    public function testDependenciesUnresolvedFeatureVariants()
```
Comment (RikudouSage) [inline tests/DefaultUnleashTest.php:829]: This test might be worth adding to client specs.

File: tests/DefaultUnleashTest.php [hunk 2]
```diff
@@ -813,6 +826,182 @@ public function testAutoRegistrationOnInvalidResponse()
...
+                    ],
+                ],
+            ],
+        ];
+        $this->pushResponse($data);
+        $instance = $this->getCachedInstance();
+        self::assertTrue($instance->isEnabled('child'));
+    }
+
+    public function testDependenciesDirectFeatureVariants()
```
Comment (RikudouSage) [inline tests/DefaultUnleashTest.php:870]: This test might be worth adding to client specs.

File: tests/DefaultUnleashTest.php [hunk 3]
```diff
@@ -813,6 +826,182 @@ public function testAutoRegistrationOnInvalidResponse()
...
+                    ],
+                ],
+            ],
+        ];
+        $this->pushResponse($data);
+        $instance = $this->getCachedInstance();
+        self::assertTrue($instance->isEnabled('child'));
+    }
+
+    public function testDependenciesMixedResolvedUnresolved()
```
Comment (RikudouSage) [inline tests/DefaultUnleashTest.php:911]: This test might be worth adding to client specs.

File: src/DefaultUnleash.php [hunk 4]
```diff
@@ -205,4 +224,35 @@ private function findStrategyHandlers(Strategy $strategy): array
 
         return $handlers;
     }
+
+    private function isDependencyEnabled(FeatureDependency $dependency, Context $context, bool $default): bool
```
Comment (kwasniew) [inline src/DefaultUnleash.php]: can we adjust domain vocabulary to `isParentDependencySatisfied`? The reason for that is that we want to emphasize that this is a parent dependency not a child dependency. Also dependency doesn't have to be enabled - the spec also defines disabled parent as an option.

Comment (RikudouSage) [reply src/DefaultUnleash.php]: Done.

File: src/Repository/DefaultUnleashRepository.php [hunk 5]
```diff
@@ -239,19 +247,161 @@ private function parseFeatures(array $body): array
...
+        $this->handleUnresolvedDependencies($unresolved, $features);
+
         return $features;
     }
 
+    /**
+     * @param array<Feature>                $unresolved
+     * @param array<string, DefaultFeature> $features
+     */
+    private function handleUnresolvedDependencies(array &$unresolved, array &$features): void
```
Comment (kwasniew) [inline src/Repository/DefaultUnleashRepository.php]: This part seems to be very different from all the other Unleash SDKs I've seen. Do we have any README/design decisions/ADRs in this repo that would explain why we're doing this part?

Comment (RikudouSage) [reply src/Repository/DefaultUnleashRepository.php]: Not really, I haven't looked at the other SDKs and this is just how I did this.

File: tests/DefaultUnleashTest.php (hunk 2 above)
Comment (kwasniew) [reply tests/DefaultUnleashTest.php:870]: I though "Child is enabled if its parent matches a specific variant." matches this case. Can you elaborate what exactly are we missing in the client spec? We'd like the spec to cover all corner cases so all the findings that you make are greatly appreciated.

Comment (RikudouSage) [reply tests/DefaultUnleashTest.php:870]: If I'm not mistaken, the spec tests only contain variants in strategies, not variants in features directly.

File: tests/DefaultUnleashTest.php (hunk 3 above)
Comment (kwasniew) [reply tests/DefaultUnleashTest.php:911]: Can you confirm if "Child is enabled if its parent matches any of the specified variants." is covering this case?

File: tests/DefaultUnleashTest.php (hunk 2 above)
Comment (kwasniew) [reply tests/DefaultUnleashTest.php:870]: That's correct since variants and strategy variants have separate spec that should prove that they behave the same. In the long term we'd like to migrate towards strategy variants and slowly sunset feature environment variants.

File: src/Repository/DefaultUnleashRepository.php [hunk 6]
```diff
@@ -239,19 +247,161 @@ private function parseFeatures(array $body): array
...
+    {
+        if (count($unresolved)) {
```
Comment (kwasniew) [inline src/Repository/DefaultUnleashRepository.php]: What do you think about extracting some smaller methods explaining the steps of the algorithm?

Just for inspiration chatgpt session suggested sth like this so take it with a grain of salt :) 
```php
    private function handleUnresolvedDependencies(array &$unresolved, array &$features): void
    {
        if (empty($unresolved)) {
            return;
        }

        do {
            $currentUnresolvedCount = count($unresolved);
            foreach ($unresolved as $index => $unresolvedItem) {
                $dependencies = $this->resolveFeatureDependencies($unresolvedItem, $features);
                $this->updateFeatureWithResolvedDependencies($unresolvedItem, $dependencies, $features);
                unset($unresolved[$index]);
            }
        } while (count($unresolved) < $currentUnresolvedCount);
    }

    private function resolveFeatureDependencies($unresolvedItem, array &$features): array
    {
        $dependencies = [];
        foreach ($unresolvedItem->getDependencies() as $dependency) {
            if (!$dependency instanceof UnresolvedFeatureDependency) {
                continue;
            }

            $feature = $this->resolveDependencyFeature($dependency, $features);
            $requiredVariants = $this->resolveDependencyVariants($dependency, $feature);

            $dependencies[] = $this->createFeatureDependency($feature, $dependency, $requiredVariants);
        }
        return $dependencies;
    }

    private function resolveDependencyFeature($dependency, array &$features)
    {
        $feature = $dependency->getFeature();
        if ($feature instanceof UnresolvedFeature) {
            return $features[$feature->getName()] ?? null;
        }
        return $feature;
    }

    private function resolveDependencyVariants($dependency, $feature): array
    {
        $requiredVariants = [];
        foreach ($dependency->getRequiredVariants() ?? [] as $requiredVariant) {
            if (!$requiredVariant instanceof UnresolvedVariant) {
                $requiredVariants[] = $requiredVariant;
                continue;
            }
            $resolvedVariant = $this->findMatchingVariant($feature, $requiredVariant);
            $requiredVariants[] = $resolvedVariant ?: $requiredVariant;
        }
        return $requiredVariants;
    }

    private function findMatchingVariant($feature, $requiredVariant)
    {
        foreach ($feature->getVariants() as $variant) {
            if ($variant->getName() === $requiredVariant->getName()) {
                return $variant;
            }
        }
        return null;
    }

    private function createFeatureDependency($feature, $dependency, array $requiredVariants): DefaultFeatureDependency
    {
        return new DefaultFeatureDependency(
            feature: $feature,
            expectedState: $dependency->getExpectedState(),
            requiredVariants: $requiredVariants,
        );
    }

    private function updateFeatureWithResolvedDependencies($unresolvedItem, array $dependencies, array &$features): void
    {
        $features[$unresolvedItem->getName()] = new DefaultFeature(
            name: $unresolvedItem->getName(),
            enabled: $unresolvedItem->isEnabled(),
            strategies: $unresolvedItem->getStrategies(),
            variants: $unresolvedItem->getVariants(),
            impressionData: $unresolvedItem->hasImpressionData(),
            dependencies: $dependencies,
        );
    }
```

Comment (RikudouSage) [reply src/Repository/DefaultUnleashRepository.php]: Sure, I can split it into smaller methods.

File: src/Repository/DefaultUnleashRepository.php [hunk 7]
```diff
@@ -239,19 +247,161 @@ private function parseFeatures(array $body): array
...
+        }
+    }
+
+    /**
+     * @param array<array{feature: string, enabled?: bool, variants?: array<string>}> $dependencies
+     * @param array<string, Feature>                                                  $features
+     *
+     * @return array<FeatureDependency>
+     */
+    private function parseDependencies(array $dependencies, array $features, ?bool &$hasUnresolvedDependencies = null): array
```
Comment (kwasniew) [inline src/Repository/DefaultUnleashRepository.php:358]: and same here, personally I'd do extract method refactoring to explain the steps of the algorithm and reduce the indentation level/whitespace on the left hand side of the code.

**Review by kwasniew — COMMENTED** [review]
I'll do final review on Monday after the extract methods refactoring.

File: src/Repository/DefaultUnleashRepository.php (hunk 6 above)
Comment (RikudouSage) [reply src/Repository/DefaultUnleashRepository.php]: I've instead rewritten the code and added comments, is that enough? Some parts of that code were from a previous version where I assumed infinite nesting of dependencies, I've removed them now.

File: src/Repository/DefaultUnleashRepository.php (hunk 7 above)
Comment (RikudouSage) [reply src/Repository/DefaultUnleashRepository.php:358]: I tried to split it up a bit, hopefully it helps.

File: src/Repository/DefaultUnleashRepository.php [hunk 8]
```diff
@@ -239,19 +247,196 @@ private function parseFeatures(array $body): array
...
         }
 
+        // $features are passed by reference so that the method can modify them
+        $this->handleUnresolvedDependencies($unresolved, $features);
+
         return $features;
     }
 
+    /**
+     * All unresolved dependencies are solved in this method (if possible).
```
Comment (kwasniew) [inline src/Repository/DefaultUnleashRepository.php:274]: This is fine for now. Every SDK will be much more complex with multi-level deps. However in our initial research we haven't found a need for this extra complexity.

**Review by kwasniew — APPROVED** [review]

File: src/Repository/DefaultUnleashRepository.php (hunk 8 above)
Comment (RikudouSage) [reply src/Repository/DefaultUnleashRepository.php:274]: Yeah, it's just the for future me or some other future implementor in case it's needed.

**Comment by kwasniew** [comment]
@RikudouSage I'll update our docs (https://docs.getunleash.io/reference/dependent-features#client-sdk-support) after you publish a new version
